File: python/Q143_reorder_list.py
#
# @lc app=leetcode id=143 lang=python3
#
# [143] Reorder List
#


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val: int = val
        self.next: ListNode | None = next


# @lc code=start
class Solution:
    def reorderList(self, head: ListNode | None) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        if head is None or head.next is None:
            return

        fast: ListNode | None = head
        slow: ListNode = head

        while fast and fast.next:
            fast = fast.next.next

            assert slow.next is not None
            slow = slow.next

        previous: ListNode | None = None
        current: ListNode | None = slow

        while current:
            tmp: ListNode | None = current.next
            current.next = previous
            previous = current
            current = tmp

        first_half: ListNode = head
        assert previous is not None
        second_half: ListNode = previous

        while True:
            tmp1: ListNode | None = first_half.next
            tmp2: ListNode | None = second_half.next

            first_half.next = second_half
            if tmp1 == slow:
                break
            second_half.next = tmp1

            assert tmp1 is not None
            assert tmp2 is not None

            first_half = tmp1
            second_half = tmp2


# @lc code=end

File: rust/src/leetcode/Q143_reorder_list.rs
/*
 * @lc app=leetcode id=143 lang=rust
 *
 * [143] Reorder List
 */

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

// @lc code=start
impl Solution {
    pub fn reorder_list(head: &mut Option<Box<ListNode>>) {
        if head.is_none() || head.as_ref().unwrap().next.is_none() {
            return;
        }

        let mut count: i32 = 0;
        let mut current: Option<&Box<ListNode>> = head.as_ref();

        while let Some(node) = current {
            count += 1;
            current = node.next.as_ref();
        }

        let mid_point: i32 = count / 2;

        let mut tail: &mut Option<Box<ListNode>> = head;

        for _ in 0..mid_point {
            tail = &mut tail.as_mut().unwrap().next;
        }

        let mut previous: Option<Box<ListNode>> = None;
        let mut current: Option<Box<ListNode>> = tail.take();

        while let Some(mut node) = current {
            current = node.next.take();
            node.next = previous;
            previous = Some(node);
        }

        let mut first_half: Option<Box<ListNode>> = head.take();
        let mut second_half: Option<Box<ListNode>> = previous;

        let mut dummy: Option<Box<ListNode>> = Some(Box::new(ListNode::new(0)));
        let mut current: &mut Option<Box<ListNode>> = &mut dummy;

        while first_half.is_some() || second_half.is_some() {
            if let Some(mut node) = first_half.take() {
                first_half = node.next.take();
                current.as_mut().unwrap().next = Some(node);
                current = &mut current.as_mut().unwrap().next;
            }

            if let Some(mut node) = second_half.take() {
                second_half = node.next.take();
                current.as_mut().unwrap().next = Some(node);
                current = &mut current.as_mut().unwrap().next;
            }
        }

        *head = dummy.unwrap().next;
    }
}
// @lc code=end
